### PR TITLE
Discovery extension

### PR DESCRIPTION
**Pull request by afuechsel**
added removeOlderThings for DiscoveryServices and a timeToLive for DiscoveryResults. HueBinding and YahooWeatherBinding extended to use removeOlderThings. UpnpDiscoveryParticipant extended to set the timeToLive.

### CODE REVIEW

File: bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/DiscoveryServiceRegistryImpl.java [hunk 1]
```diff
@@ -221,6 +222,22 @@ public synchronized void thingRemoved(DiscoveryService source, ThingUID thingUID
         }
     }
 
+    @Override
+    public Collection<ThingUID> removeOlderResults(DiscoveryService source, long timestamp,
+            Collection<ThingTypeUID> thingTypeUIDs) {
+        HashSet<ThingUID> removedThings = new HashSet<>();
```
Comment (kaikreuzer) [inline bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/DiscoveryServiceRegistryImpl.java]: should probably also be renamed to removedResults

File: bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/DiscoveryResultImpl.java [hunk 2]
```diff
@@ -215,7 +228,16 @@ public boolean equals(Object obj) {
     @Override
     public String toString() {
         return "DiscoveryResult [thingUID=" + thingUID + ", properties=" + properties + ", flag=" + flag + ", label="
-                + label + ", bridgeUID=" + bridgeUID + "]";
+                + label + ", bridgeUID=" + bridgeUID + ", ttl=" + timeToLive + ", createdAt=" + createdAt + "]";
+    }
+    
+    @Override
+    public long getCreatedAt() {
```
Comment (kaikreuzer) [inline bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/DiscoveryResultImpl.java]: "created" is probably not fully correct - isn't this also updated, if a result entry is updated?

File: bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/AbstractDiscoveryService.java [hunk 3]
```diff
@@ -33,13 +36,12 @@
  * @author Oliver Libutzki - Initial contribution
  * @author Kai Kreuzer - Refactored API
  * @author Dennis Nobel - Added background discovery configuration through Configuration Admin
+ * @author Andre Fuechsel - Added removeOlderResults
  */
 public abstract class AbstractDiscoveryService implements DiscoveryService {
 
     private final Logger logger = LoggerFactory.getLogger(AbstractDiscoveryService.class);
 
-    static protected final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(3);
```
Comment (kaikreuzer) [inline bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/AbstractDiscoveryService.java]: In order to make it easy for users (and not to break backward compatibility), I would suggest to keep the scheduler field here and just initialize it with ```DiscoveryThreadPool.getScheduler()```

File: bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/DiscoveryResultImpl.java (hunk 2 above)
Comment (afuechsel) [inline bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/DiscoveryResultImpl.java]: Originally I named it "timestamp". I changed it on Dennis' suggestion. I have the feeling, that I was not so wrong at all.... Should I rename it to timestamp?

Comment (dnobel) [inline bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/DiscoveryResultImpl.java]: Well, naming a variable "timestamp" would be the same as naming it "integer" or "boolean" or "string". It says nothing about the meaning of this variable. So what kind of timestamp is it? This was the reason why I suggested to give it another name. I know it is already merged, but we should avoid to name public methods or variables like this.

**Pull request merged**